### PR TITLE
Add the hybrid cryptosystem support

### PR DESCRIPTION
**Pull request by ezimuel**
This PR adds the [hybrid cryptosystem](https://en.wikipedia.org/wiki/Hybrid_cryptosystem) support (OpenPGP like). It allows also multiple encryptions using a keyrings of public/private keys. This can be useful in multi users scenarios to encrypt data only for specific users.

### CODE REVIEW

**Comment by ezimuel** [comment]
@paragonie-scott do you have time for a quick review? Thanks!

File: doc/book/hybrid.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
+# Encrypt and decrypt using hybrid cryptosystem
+
```
Comment (weierophinney) [inline doc/book/hybrid.md:2]: Add a note here along the lines of `- Since 3.1.0`

File: README.md [hunk 2]
```diff
@@ -4,11 +4,12 @@
 [![Coverage Status](https://coveralls.io/repos/zendframework/zend-crypt/badge.svg?branch=master)](https://coveralls.io/r/zendframework/zend-crypt?branch=master)
 
 `Zend\Crypt` provides support of some cryptographic tools.
-The available features are:
+Some of the available features are:
 
 - encrypt-then-authenticate using symmetric ciphers (the authentication step
   is provided using HMAC);
 - encrypt/decrypt using symmetric and public key algorithm (e.g. RSA algorithm);
+- encrypt/decrypt using hybrid mode (OpenPHP like);
```
Comment (paragonie-scott) [inline README.md]: OpenPGP?

**Comment by paragonie-scott** [comment]
This looks good. It's essentially what [EasyRSA](https://github.com/paragonie/EasyRSA) tries to provide.

Since this is a new feature, I'd definitely advise forcing the use of OAEP.

**Comment by ezimuel** [comment]
@paragonie-scott thanks for your feedback! You right, I always forget OAEP :)

File: doc/book/hybrid.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+Hybrid is an encryption mode that uses symmetric and public keys ciphers together.
```
Comment (weierophinney) [inline doc/book/hybrid.md]: `s/keys/key/`

File: doc/book/hybrid.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+The idea is to take the advantages of the public key cryptography for sharing the
+keys and the speed of symmmetric encryption to encrypt the message.
```
Comment (weierophinney) [inline doc/book/hybrid.md]: Rephrase to:

```
The approach takes advantage of public key cryptography for sharing keys and symmetric encryption speed for encrypting messages.
```

File: doc/book/hybrid.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+The hybrid mode is able to encrypt message for one or more receivers and can be
+used in multi user scenario, where you can limit the decryption only for some users.
```
Comment (weierophinney) [inline doc/book/hybrid.md]: Rephrase to:

```
Hybrid mode allows you to encrypt a message for one or more receivers, and can be used in multi-user scenarios where you wish to limit decryption to specific users.
```

File: doc/book/hybrid.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+## How it works
+
+Suppose we have two users: *Alice* and *Bob*. *Alice* wants to send a message to *Bob*
+using an hybrid cryptosystem, she needs to:
```
Comment (weierophinney) [inline doc/book/hybrid.md]: `s/an/a/`

File: doc/book/hybrid.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+- Obtain *Bob*'s public key;
+- Generates a random session key (one-time pad);
+- Encrypts the message using a symmetric cipher with the previous session key;
+- Encrypts the session key using the *Bob*'s public key;
```
Comment (weierophinney) [inline doc/book/hybrid.md]: `s/the//`

File: doc/book/hybrid.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+- Send both of these encryptions to *Bob*.
```
Comment (weierophinney) [inline doc/book/hybrid.md]: Rephrase to `Sends both the encrypted message and encrypted session key to *Bob*.`

File: doc/book/hybrid.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+A schema of the encryption is reported in the image below:
+
+![Encryption schema](images/zend.crypt.hybrid.png)
+
+To decrypt the message *Bob* needs to:
```
Comment (weierophinney) [inline doc/book/hybrid.md]: `s/message/message,/`

File: doc/book/hybrid.md [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+- Uses his private key to decrypt the session key;
+- Uses this session key to decrypt the message.
+
+## Example of usage
+
+In order to use the `Zend\Crypt\Hybrid` component you need to have a keyring of
```
Comment (weierophinney) [inline doc/book/hybrid.md]: `s/component/component,/`

File: doc/book/hybrid.md [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+public and private keys. To encrypt a message you can use the following code:
```
Comment (weierophinney) [inline doc/book/hybrid.md]: Rephrase to `To encrypt a message, use the following code:`

File: doc/book/hybrid.md [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+$hybrid     = new Hybrid();
+$ciphertext = $hybrid->encrypt('message', $publicKey);
+$plaintext  = $hybrid->decrypt($ciphertext, $privateKey);
+
+printf($plaintext === 'message' ? "Success\n" : "Error\n");
+```
+
+We generated the keys using the [Zend\Crypt\PublicKey\RsaOptions](public-key.md)
+component. You can also use a [PEM](https://en.wikipedia.org/wiki/Privacy-enhanced_Electronic_Mail)
+strings for the keys. If you use a string for the private key you need to pass
```
Comment (weierophinney) [inline doc/book/hybrid.md]: - `s/strings/string/`
- `s/key/key,/`

File: doc/book/hybrid.md [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+the pass-phrase for decrypt, if present, like in the following example:
```
Comment (weierophinney) [inline doc/book/hybrid.md]: Rephrase to `the pass phrase to use when decrypting`.

File: doc/book/hybrid.md [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+$privateKey = $rsaOptions->getPrivateKey()->toString();
+
+$hybrid     = new Hybrid();
+$ciphertext = $hybrid->encrypt('message', $publicKey);
+$plaintext  = $hybrid->decrypt($ciphertext, $privateKey, 'test'); // pass-phrase
+
+printf($plaintext === 'message' ? "Success\n" : "Error\n");
+```
+
+The `Hybrid` component uses the `Zend\Crypt\BlockCipher` for the symmetric
```
Comment (weierophinney) [inline doc/book/hybrid.md]: Remove the `the` before the `BlockCipher` class reference.

File: doc/book/hybrid.md [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+cipher and the `Zend\Crypt\Rsa` for the public-key cipher.
```
Comment (weierophinney) [inline doc/book/hybrid.md]: Same here with `Rsa`.

File: doc/book/hybrid.md [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+## Encrypt with multiple keys
+
+The `Zend\Crypt\Hybrid` component can be used to encrypt a message for multiple
+users, using a keyring of Ids and public keys. This keyring can be specified using
```
Comment (weierophinney) [inline doc/book/hybrid.md]: `s/Ids/identifiers/`

File: doc/book/hybrid.md [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+an array of `[ 'id' => 'publickey' ]`, where `publickey` can be a string (PEM)
+or an instance of `Zend\Crypt\PublicKey\Rsa\PublicKey`. The `id` can be any
+string, for instance the email address of the users.
```
Comment (weierophinney) [inline doc/book/hybrid.md]: `s/,.*$/; for example, a receipient email address./`

File: doc/book/hybrid.md [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+
+Here is reported an example of encryption using a keyring of 4 keys:
```
Comment (weierophinney) [inline doc/book/hybrid.md]: Rephrase to: `The following details encryption using a keyring with 4 keys:`

File: mkdocs.yml [hunk 19]
```diff
@@ -6,6 +6,7 @@ pages:
     - Reference:
       - { 'Block Ciphers': block-cipher.md }
       - { 'Encrypting Files': files.md }
+      - { 'Hybrid cryptosystem': hybrid.md }
```
Comment (weierophinney) [inline mkdocs.yml]: Capitalize `Cryptosystem`.

File: src/Hybrid.php [hunk 20]
```diff
@@ -0,0 +1,143 @@
+<?php
+namespace Zend\Crypt;
```
Comment (weierophinney) [inline src/Hybrid.php:10]: Please add a copyright/license file-level docblock.

File: src/Hybrid.php [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+use Zend\Crypt\PublicKey\Rsa\PublicKey as PubKey;
+use Zend\Crypt\PublicKey\Rsa\PrivateKey;
+
+/**
+ * Hybrid encryption (OpenPGP like)
+ *
+ * The data are encrypted using a BlockCipher with a random session key
+ * that is encrypted using RSA with the public key of the receiver.
+ * The decryption process retrieves the session key using RSA with the private
+ * key of the receiver and decrypt the data using the BlockCipher.
```
Comment (weierophinney) [inline src/Hybrid.php]: `s/decrypt/decrypts/`

File: src/Hybrid.php [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+    public function encrypt($plaintext, $keys = null)
+    {
+        // generate a random session key
+        $sessionKey = Rand::getBytes($this->bCipher->getCipher()->getKeySize());
+
+        // encrypt the plaintext with blockcipher algorithm
+        $this->bCipher->setKey($sessionKey);
+        $ciphertext = $this->bCipher->encrypt($plaintext);
+
+        if (!is_array($keys)) {
```
Comment (weierophinney) [inline src/Hybrid.php]: Space following the `!`, please (you do it that way elsewhere in the class).

File: src/Hybrid.php [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+            $keys = [ '' => $keys ];
```
Comment (weierophinney) [inline src/Hybrid.php]: No space following the `[` nor before the `]`, please.

File: src/Hybrid.php [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+            if (! $pubkey instanceof PubKey) {
+                if (is_string($pubkey)) {
+                    $pubkey = new PubKey($pubkey);
+                } else {
+                    throw new Exception\RuntimeException(sprintf(
+                        "The public key must be a string in PEM format or an instance of %s",
+                        PubKey::class
+                    ));
+                }
+            }
```
Comment (weierophinney) [inline src/Hybrid.php:77]: Let's re-flow this to:

``` php
if (! $pubkey instanceof PubKey && ! is_string($pubkey)) {
    throw new Exception\RuntimeException(sprintf(
        'The public key must be a string in PEM format or an instance of %s',
        PubKey::class
    ));
}

$pubkey = is_string($pubkey) ? new PubKey($pubkey) : $pubkey;
```

File: src/Hybrid.php [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+                "%s:%s:",
+                base64_encode($id),
+                base64_encode($this->rsa->encrypt($sessionKey, $pubkey))
+            );
+        }
+        return $encKeys . ';' . $ciphertext;
+    }
+
+    /**
+     * Decrypt usign a private key
```
Comment (weierophinney) [inline src/Hybrid.php]: `s/usign/using/`

File: src/Hybrid.php [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+        if (! $privateKey instanceof PrivateKey) {
+            if (is_string($privateKey)) {
+                $privateKey = new PrivateKey($privateKey, $passPhrase);
+            } else {
+                throw new Exception\RuntimeException(sprintf(
+                    "The private key must be a string in PEM format or an instance of %s",
+                    PrivateKey::class
+                ));
+            }
+        }
```
Comment (weierophinney) [inline src/Hybrid.php:116]: Use the same structure as I proposed above for this if/else block.

File: test/HybridTest.php [hunk 27]
```diff
@@ -0,0 +1,175 @@
...
+namespace ZendTest\Crypt;
+
+use Zend\Crypt\Hybrid;
+use Zend\Crypt\BlockCipher;
+use Zend\Crypt\PublicKey\Rsa;
+use Zend\Crypt\PublicKey\RsaOptions;
+
+/**
+ * @group      Zend_Crypt
+ */
```
Comment (weierophinney) [inline test/HybridTest.php]: These annotations aren't necessary any more.

File: test/HybridTest.php [hunk 28]
```diff
@@ -0,0 +1,175 @@
...
+class HybridTest extends \PHPUnit_Framework_TestCase
+{
+    protected $hybrid;
+
+    public function setUp()
+    {
+        if (!extension_loaded('openssl')) {
```
Comment (weierophinney) [inline test/HybridTest.php]: Please add a space following the `!` operator.

**Comment by weierophinney** [comment]
Overall, relatively straight-forward, and the code was mostly easy to follow. Nice feature, @ezimuel !

**Comment by ezimuel** [comment]
@paragonie-scott Btw, we already use OAEP as default padding for encryption using RSA, since zend-crypt 2.4.9 as reported in [ZF2015-10](http://framework.zend.com/security/advisory/ZF2015-10).

**Comment by ezimuel** [comment]
@weierophinney I included all the proposed changes. It's ready to be merged, thanks!

File: .gitignore [hunk 29]
```diff
@@ -11,6 +11,7 @@ tmp/
 zf-mkdoc-theme/
 
 clover.xml
+composer.lock
```
Comment (weierophinney) [inline .gitignore:14]: This should be removed; the travis setup uses the lowest/locked/latest strategy, so the file needs to be present. I'll re-add it on merge.

**Comment by weierophinney** [comment]
Thanks for the great feature, @ezimuel!

**Pull request merged**